### PR TITLE
GraphQl-93: Implement support for variables in query

### PR DESCRIPTION
**Pull request by Aquarvin**
-- Variables may be Input type
-- Query now accepts variables
-- added type property for Output/Input Type element
-- functional test was added

### Description (*)
Currently running the following Graph Ql query
```
query GetProductsQuery($page: Int, $filterInput: ProductFilterInput){
  products(
    pageSize: 10
    currentPage: $page
    filter: $filterInput
    sort: {}
  ) {
    items {
    	name
    }
  }
}
```
with variables:
```
{
  "page": 1,
  "filterInput": {
    "price": {
    	"gt": "10"
    }
  }
}
```
Will return filtered products list, since ProductFilterInput is an input type, according to scheme
### Fixed Issues (if relevant)
1. magento/graphql-ce#93: Implement support for variables in query

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [+ ] Pull request has a meaningful description of its purpose
 - [ +] All commits are accompanied by meaningful commit messages
 - [ +] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by Aquarvin** [comment]
Request Sample:
```
query GetProductsQuery($page: Int, $filterInput: ProductFilterInput){
  products(
    pageSize: 10
    currentPage: $page
    filter: $filterInput
    sort: {}
  ) {
    items {
    	name
    }
  }
}
```
With variables:
```
{
  "page": 1,
  "filterInput": {
    "price": {
    	"gt": "10"
    }
  }
}
```

**Review by paliarush — APPROVED** [review]

File: dev/tests/api-functional/testsuite/Magento/GraphQl/VariablesSupportQueryTest.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+     * @magentoApiDataFixture Magento/Catalog/_files/product_simple_with_all_fields.php
+     * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
+     */
+    public function testQueryObjectVariablesSupport()
+    {
+        $productSku = 'simple';
+
+        $query
+            = <<<'QUERY'
+query GetProductsQuery($page: Int, $filterInput: ProductFilterInput){
```
Comment (paliarush) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/VariablesSupportQueryTest.php]: Do we have enum used here?

File: lib/internal/Magento/Framework/GraphQl/Config.php [hunk 2]
```diff
@@ -81,18 +81,24 @@ public function getConfigElement(string $configElementName) : ConfigElementInter
     }
 
     /**
-     * Return all type names declared in a GraphQL schema's configuration.
+     * Return all type names declared in a GraphQL schema's configuration and their type.
      *
-     * @return string[]
+     * @return array $types
+     *  name string
```
Comment (paliarush) [inline lib/internal/Magento/Framework/GraphQl/Config.php]: Is this common style for documenting array structure? Usually I see 
`['name' => 'example value', 'type' = 'example value']`

File: lib/internal/Magento/Framework/GraphQl/Config.php [hunk 3]
```diff
@@ -81,18 +81,24 @@ public function getConfigElement(string $configElementName) : ConfigElementInter
...
+     *  type string
      */
     public function getDeclaredTypeNames() : array
```
Comment (paliarush) [inline lib/internal/Magento/Framework/GraphQl/Config.php]: Can we rename this method, to `getDeclaredTypes`, for example

File: lib/internal/Magento/Framework/GraphQl/Config/Element/InputFactory.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            );
+        }
+        return $this->create(
+            $data,
+            $fields
+        );
+    }
+
+    /**
+     * Create type object based off array of configured GraphQL InputType data.
```
Comment (paliarush) [inline lib/internal/Magento/Framework/GraphQl/Config/Element/InputFactory.php]:
```suggestion
     * Create input type object based off array of configured GraphQL InputType data.
```

File: lib/internal/Magento/Framework/GraphQl/Config/Element/Type.php [hunk 5]
```diff
@@ -8,7 +8,7 @@
 namespace Magento\Framework\GraphQl\Config\Element;
 
 /**
- * Describes all the configured data of an Output or Input type in GraphQL.
+ * Class representing 'type' GraphQL config element.
```
Comment (paliarush) [inline lib/internal/Magento/Framework/GraphQl/Config/Element/Type.php:11]:
```suggestion
 * Class representing 'type' GraphQL config elements.
```

File: lib/internal/Magento/Framework/GraphQl/Query/Fields.php [hunk 6]
```diff
@@ -62,4 +67,24 @@ public function getFieldsUsedInQuery()
...
     }
+
+    /**
+     * Extract and return list of all used fields in GraphQL query's variables
+     *
+     * @param array $variables
+     *
+     * @return string[]
+     */
+    private function getVariables(array $variables): array
```
Comment (paliarush) [inline lib/internal/Magento/Framework/GraphQl/Query/Fields.php]:
```suggestion
    private function formatVariables(array $variables): array
```
or maybe `extractVariables`

File: lib/internal/Magento/Framework/GraphQl/Schema/SchemaGenerator.php [hunk 7]
```diff
@@ -55,16 +54,15 @@ public function generate() : Schema
...
                 'typeLoader' => function ($name) {
-                    return $this->outputMapper->getOutputType($name);
+                    return $this->typeRegistry->get($name);
                 },
                 'types' => function () {
-                    //all types should be generated only on introspection
                     $typesImplementors = [];
-                    foreach ($this->config->getDeclaredTypeNames() as $name) {
-                        $typesImplementors [] = $this->outputMapper->getOutputType($name);
+                    foreach ($this->config->getDeclaredTypeNames() as $type) {
```
Comment (paliarush) [inline lib/internal/Magento/Framework/GraphQl/Schema/SchemaGenerator.php]: Please create a ticket not to load all types, but just the ones which do not have references from Query/Mutation types

File: lib/internal/Magento/Framework/GraphQl/Schema/Type/TypeRegistry.php [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+     */
+    public function get(string $typeName): TypeInterface
+    {
+        if (!isset($this->types[$typeName])) {
+            $configElement = $this->config->getConfigElement($typeName);
+
+            $configElementClass = get_class($configElement);
+            if (!isset($this->configToTypeMap[$configElementClass])) {
+                throw new GraphQlInputException(
+                    new Phrase("Type for '{$configElementClass}' has not declared.")
```
Comment (paliarush) [inline lib/internal/Magento/Framework/GraphQl/Schema/Type/TypeRegistry.php]:
```suggestion
                    new Phrase("No mapping to Webonyx type is declared for '{$configElementClass}' config element type.")
```

**Review by paliarush — CHANGES_REQUESTED** [review]

**Comment by magento-engcom-team** [comment]
@Aquarvin thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by danslo** [comment]
I'm not sure if this fixes every case. For example, I try with this query:

```
query testQuery($sort: ProductSortInput) {
  products(sort: $sort) {
    items {
      name
    }
  }
}
```

With query variables:
```
{
  "sort": {
    "name":"ASC"
  }
}
```

Getting these errors:
- "Variable \"$sort\" cannot be non-input type \"ProductSortInput\".",
- "Variable \"$sort\" of type \"ProductSortInput\" used in position expecting type \"ProductSortInput\".",

**Pull request merged**